src/Crypto/Encoding/BIP39.hs

```diff
@@ -290,3 +302,135 @@ phraseToSeed mw dic passphrase =
   where
     sentence = toData $ mnemonicPhraseToString dic mw
     toData = String.toBytes String.UTF8
+
+
+-- -------------------------------------------------------------------------- --
+-- Mnemonic Sentence and Mnemonic Phrase
+-- -------------------------------------------------------------------------- --
+
+-- | Mnemonic Sentence is a list of 'WordIndex'.
+--
+-- This is the generic representation of a mnemonic phrase that can be used for
+-- transalating to a different dictionary (example: English to Japanese).
+--
+-- This is mainly used to convert from/to the 'Entropy' and for 'cardanoSlSeed'
+--
+newtype MnemonicSentence (mw :: Nat) = MnemonicSentence
+    { mnemonicSentenceToListN :: ListN mw WordIndex
+    }
+  deriving (Show, Eq, Ord, Typeable, NormalForm)
+instance ValidMnemonicSentence mw => IsList (MnemonicSentence mw) where
+    type Item (MnemonicSentence mw) = WordIndex
+    fromList = MnemonicSentence . fromMaybe (error "invalid mnemonic size") . ListN.toListN
+    toList = ListN.unListN . mnemonicSentenceToListN
+
+-- | Type Constraint to validate the given 'Nat' is valid for the supported
+-- 'MnemonicSentence'
+type ValidMnemonicSentence (mw :: Nat) =
+    ( KnownNat mw
+    , NatWithinBound Int mw
+    , Elem mw '[9, 12, 15, 18, 21, 24]
+    )
+
+-- | Human readable representation of a 'MnemonicSentence'
+--
+newtype MnemonicPhrase (mw :: Nat) = MnemonicPhrase
+    { mnemonicPhraseToListN :: ListN mw String
+    }
+  deriving (Show, Eq, Ord, Typeable, NormalForm)
+instance ValidMnemonicSentence mw => IsList (MnemonicPhrase mw) where
+    type Item (MnemonicPhrase mw) = String
+    fromList = fromRight (error "invalid mnemonic phrase") . mnemonicPhrase
+    toList = ListN.unListN . mnemonicPhraseToListN
+
+mnemonicPhrase :: forall mw csz. (ValidMnemonicSentence mw)
+               => [String] -> Either (EntropyError csz) (MnemonicPhrase mw)
+mnemonicPhrase l = maybe
+    (Left $ ErrInvalidEntropyLength (length l) (fromInteger $ natVal $ Proxy @mw))
```

Comment:
Hm, I believe this is wrong. It's using the number of mnemonic words instead of the entropy size.

Reply:
Yup. That doesn't really make sense to use the `ErrInvalidEntropyLength` here. Better define a new `ErrWrongNumberOfWords (actual :: Int) (expected :: Int)` error here :+1:

Reply:
To be clear: I am talking about a new type `NumberOfWordsError` and not another branch for the `EntropyError` ADT :+1: